### ISSUE TITLE
Field type falsely reported as unknown with `dataclasses.field`

### ISSUE THREAD

**Issue by aDotInTheVoid** [microsoft/pyright#10403]
Code sample in [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZIFABmUA9gLYgAmAhkldgDZXroCm6IMpYxUSlNdRswA0IAjBb0KePAAFqtBk3R4lzEAGEUeELpCkI9GJjRHMAbRgA7JAF0QAXjESpACgosCVAK70kAfS9cHggHMyQASjwgA)

```python
# pyright: strict
from dataclasses import dataclass, field

@dataclass
class C:
    mylist: list[int] = field(default_factory=list)

```

This errors with:

```
/Users/aloenr01/x.py
  /Users/aloenr01/x.py:6:5 - error: Type of "mylist" is partially unknown
    Type of "mylist" is "list[Unknown]" (reportUnknownVariableType)
1 error, 0 warnings, 0 informations 
```

Even though `mylist` is given a type (i.e `list[int]`)

**VS Code extension or command-line**

pyright 1.1.400 cli.

Interestingly, this worked in pyright 1.1.398, but broke in 1.1.399:

```
% uvx pyright@1.1.398 ./x.py
Installed 3 packages in 89ms
WARNING: there is a new pyright version available (v1.1.398 -> v1.1.400).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

0 errors, 0 warnings, 0 informations 
% uvx pyright@1.1.399 ./x.py
Installed 3 packages in 79ms
WARNING: there is a new pyright version available (v1.1.399 -> v1.1.400).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

/Users/aloenr01/x.py
  /Users/aloenr01/x.py:6:5 - error: Type of "mylist" is partially unknown
    Type of "mylist" is "list[Unknown]" (reportUnknownVariableType)
1 error, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Duplicate of #10283.
